packages/web/src/features/translation/TranslationView.tsx

```diff
@@ -1,42 +1,137 @@
-import { LoaderFunctionArgs, useLoaderData } from 'react-router-dom';
+import { useParams } from 'react-router-dom';
+import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query';
 import apiClient from '../../shared/apiClient';
 import { useLayoutContext } from '../../app/Layout';
-import {
-  GetVerseGlossesResponseBody,
-  GetVerseResponseBody,
-} from '@translation/api-types';
-
-export async function translationViewLoader({ params }: LoaderFunctionArgs) {
-  const verse = await apiClient.verses.findById(params.verseId ?? 'unknown');
-  const referenceGlosses = await apiClient.verses.findVerseGlosses(
-    params.verseId ?? 'unknown',
-    'en'
+import TextInput from '../../shared/components/TextInput';
+import { useState } from 'react';
+import InputHelpText from '../../shared/components/InputHelpText';
+import { GetVerseGlossesResponseBody } from '@translation/api-types';
+import { Icon } from '../../shared/components/Icon';
+
+export default function TranslationView() {
+  const params = useParams() as { verseId: string };
+  const { language } = useLayoutContext();
+
+  const verseQuery = useQuery(['verse', params.verseId], () =>
+    apiClient.verses.findById(params.verseId)
+  );
+  const referenceGlossesQuery = useQuery(
+    ['verse-glosses', 'en', params.verseId],
+    () => apiClient.verses.findVerseGlosses(params.verseId, 'en')
+  );
+  const targetGlossesQuery = useQuery(
+    ['verse-glosses', language, params.verseId],
+    () => apiClient.verses.findVerseGlosses(params.verseId, language)
   );
 
-  return { verse, referenceGlosses };
-}
+  const [glossRequests, setGlossRequests] = useState<
+    { wordId: string; requestId: number }[]
+  >([]);
+  const queryClient = useQueryClient();
+  const glossMutation = useMutation({
+    mutationFn: (variables: { wordId: string; gloss: string }) =>
+      apiClient.words.updateGloss(variables.wordId, language, variables.gloss),
+    onMutate: ({ wordId }) => {
+      const requestId = Math.floor(Math.random() * 1000000);
+      setGlossRequests((requests) => [...requests, { wordId, requestId }]);
+      return { requestId };
+    },
+    onError: () => {
+      alert('Unknown error occurred.');
+    },
+    onSuccess: (_, { wordId, gloss }, context) => {
+      if (context?.requestId) {
+        setGlossRequests((requests) =>
+          requests.filter((r) => r.requestId !== context.requestId)
+        );
+      }
```

Comment:
Then on updating a gloss, we have to keep track of which requests are in progress to show a loading spinner under the word.